Sources/Controllers/2.0/OASelectSubcategoryViewController.mm

```diff
@@ -7,214 +7,285 @@
 {
-    OAMultiselectableHeaderView *_headerView;
-    NSArray<NSString *> *_keys;
-    NSArray<NSString *> *_data;
     OAPOICategory *_category;
-    NSSet<NSString *> *_subcategories;
-    BOOL _selectAll;
+    OAPOIUIFilter *_filter;
+    NSMutableArray *_items;
+    NSMutableArray *_selectedItems;
 }
 
-- (instancetype)initWithCategory:(OAPOICategory *)category subcategories:(NSSet<NSString *> *)subcategories selectAll:(BOOL)selectAll
+- (instancetype)initWithCategory:(OAPOICategory *)category filter:(OAPOIUIFilter *)filter
 {
     self = [super init];
     if (self)
     {
         _category = category;
-        _subcategories = subcategories;
-        _selectAll = selectAll;
+        _filter = filter;
         [self initData];
     }
     return self;
 }
 
-- (void) initData
+- (void)initData
 {
     if (_category)
     {
-        OAPOIHelper *helper = [OAPOIHelper sharedInstance];
-        NSMutableDictionary<NSString *, NSString *> *subMap = [NSMutableDictionary dictionary];
-        for (NSString *name in _subcategories)
-            [subMap setObject:[helper getPhraseByName:name] forKey:name];
-
-        for (OAPOIType *pt in _category.poiTypes)
-            [subMap setObject:pt.nameLocalized forKey:pt.name];
-        
-        NSMutableArray<NSString *> *keys = [NSMutableArray arrayWithArray:subMap.allKeys];
-        NSMutableArray<NSString *> *data = [NSMutableArray arrayWithArray:subMap.allValues];
-        
-        [keys sortUsingComparator:^NSComparisonResult(NSString * _Nonnull name1, NSString * _Nonnull name2) {
-            return [[subMap objectForKey:name1] localizedCaseInsensitiveCompare:[subMap objectForKey:name2]];
-        }];
-        [data sortUsingComparator:^NSComparisonResult(NSString * _Nonnull nameLoc1, NSString * _Nonnull nameLoc2) {
-            return [nameLoc1 localizedCaseInsensitiveCompare:nameLoc2];
-        }];
-
-        _data = [NSArray arrayWithArray:data];
-        _keys = [NSArray arrayWithArray:keys];
+        NSSet<NSString *> *acceptedTypes = [[_filter getAcceptedTypes] objectForKey:_category];
+        NSSet<NSString *> *acceptedSubtypes = [_filter getAcceptedSubtypes:_category];
+        NSArray<OAPOIType *> *types = _category.poiTypes;
+
+        _selectedItems = [NSMutableArray new];
+        _items = [NSMutableArray arrayWithArray:[types sortedArrayUsingComparator:^NSComparisonResult(OAPOIType * _Nonnull t1, OAPOIType * _Nonnull t2) {
+            return [t1.nameLocalized localizedCaseInsensitiveCompare:t2.nameLocalized];
+        }]];
+
+        if (acceptedSubtypes == [OAPOIBaseType nullSet] || acceptedTypes.count == types.count)
+            _selectedItems = [NSMutableArray arrayWithArray:_items];
+        else
+            for (OAPOIType *poiType in _items)
+                if ([acceptedTypes containsObject:poiType.name])
+                    [_selectedItems addObject:poiType];
+
+        [_items insertObject:[[OACustomSelectionButtonCell alloc] init] atIndex:0];
```

Comment:
Never create the cells yourself, this defeats the purpose of using the tableView. All the cells should be allocated properly in the cellForRow method.